kubespawner/spawner.py

```diff
@@ -227,25 +203,46 @@ def __init__(self, *args, **kwargs):
         # The attribute needs to exist, even though it is unset to start with
         self._start_future = None
 
-    def _set_k8s_client_configuration(self):
+    @classmethod
+    async def create(cls, *args, **kwargs):
+        """In an ideal world, you'd get a new KubeSpawner with this.
+        That's not how we are called from JupyterHub, though; it just
+        instantiates whatever class you tell it to use as the spawner class.
+
+        Thus we need to do the initialization in poll(), start(), and stop()
+        since any of them could be the first thing called (imagine that the
+        hub restarts and the first thing someone does is try to stop their
+        running server).  Further a pre-spawn-start hook that depends on
+        internal knowledge of the spawner might be the first thing to run.
+
+        It would be nice if there were an agreed-upon mechanism for, say,
+        JupyterHub to look for an async create() method and then use
+        that (instead of implicitly requesting __init__()) to get its
+        spawner instance.
+        """
+        inst = cls(*args, **kwargs)
+        await inst.initialize_reflectors_and_clients()
+        return inst
+
+    async def initialize_reflectors_and_clients(self):
+        await load_config()
         # The actual (singleton) Kubernetes client will be created
-        # in clients.py shared_client but the configuration
+        # in clients.py shared_client() but the configuration
         # for token / ca_cert / k8s api host is set globally
-        # in kubernetes.py syntax.  It is being set here
-        # and this method called prior to shared_client
-        # for readability / coupling with traitlets values
-        try:
-            kubernetes.config.load_incluster_config()
-        except kubernetes.config.ConfigException:
-            kubernetes.config.load_kube_config()
+        # in kubernetes.py syntax.  It is being set here (but after
+        # load_config()) for readability / coupling with traitlets values
         if self.k8s_api_ssl_ca_cert:
             global_conf = client.Configuration.get_default_copy()
             global_conf.ssl_ca_cert = self.k8s_api_ssl_ca_cert
             client.Configuration.set_default(global_conf)
         if self.k8s_api_host:
             global_conf = client.Configuration.get_default_copy()
-            global_conf.host = self.k8s_api_host
+            global_conf.ssl_ca_cert = self.k8s_api_ssl_ca_cert
```

Comment:
```suggestion
            global_conf.host = self.k8s_api_host
```

Reply:
I implemented this by adding an optional "caller" parameter which users of load_config() can fill out with a reference to themselves; in that case load_config checks whether the caller has the right attributes, and whether those attributes have values.

I'm not wild about that, but I am even less wild about implementing a mixin class with only one method that Reflector, Proxy, and Spawner all inherit from, which was my other idea.